src/api_midlevel.jl

```diff
@@ -0,0 +1,46 @@
+# This file defines midlevel api wrappers. We include name normalization for methods that are
+# applicable to different hdf5 api-layers. We still try to adhere close proximity to the underlying
+# method name in the hdf5-library.
+
+"""
+    HDF5.set_extent_dims(dset::HDF5.Dataset, new_dims::Dims)
+
+Change the current dimensions of a dataset to `new_dims`, limited by
+`max_dims = get_simple_extent_dims(dset)[2]`. Reduction is possible and leads to loss of truncated data.
+"""
+function set_extent_dims(dset::Dataset, size::Dims)
+    checkvalid(dset)
+    h5d_set_extent(dset, hsize_t[reverse(size)...])
+end
+
+"""
+    HDF5.set_extent_dims(dset::HDF5.Dataspace, new_dims::Dims, max_dims::Union{Dims,Nothing} = nothing)
+
+Change the current dimensions of a dataspace `dset` to `new_dims`, limited by
+`max_dims = get_simple_extent_dims(dset)[2]`. Reduction is possible and leads to loss of truncated data.
+"""
+function set_extent_dims(dspace::Dataspace, size::Dims, max_dims::Union{Dims,Nothing} = nothing)
+    checkvalid(dspace)
+    rank = length(size)
+    current_size = hsize_t[reverse(size)...]
+    maximum_size = isnothing(max_dims) ? C_NULL : hsize_t[reverse(max_dims)...]

```

Comment:
```suggestion
    maximum_size = isnothing(max_dims) ? C_NULL : [reverse(max_dims .% hsize_t)...]
```
to support using `-1` to set unlimited max size.

(Unfortunately, using `max_dims::Dims` makes this incompatible with unlimited sizes given as `(HDF5.H5S_UNLIMITED,)` ... not sure what to do about that, but it's not a regression from what `set_dims!` was doing already.)

Reply:
Yeah I think this will take some more careful thinking on how to address  setting `H5S_UNLIMITED` more broadly.

Reply:
I think the simplest thing is to just document the correct way to set an unlimited dimension is to set it to `-1`  and not use the internal HDF5.H5S_UNLIMITED constant?  A little more instrusive would be to set `HDF5.H5S_UNLIMITED=-1`, but I think that would require enforcing truncation to which ever inter api call excepts `HDF5.H5S_UNLIMITED`  since `cconvert` to `hsize_t` would throw an error.

Reply:
I agree that it's mostly a documentation issue — low-level API will want `HDF5.H5S_UNLIMITED` (and should remain as `-1 % hsize_t` to agree with the C header literal value), and everything in mid-level and high-level API will instead use `-1` and do the data type conversions. I think so far we just don't have the documentation to make that clear.